controller/uninstall_controller.go

```diff
@@ -0,0 +1,437 @@
+				logrus.WithFields(logFields).Warnf("Failed to mark for deletion")
+				failed = true
+			} else {
+				logrus.WithFields(logFields).Infof("Marked for deletion")
+			}
+		} else if vol.DeletionTimestamp.Before(&timeout) {
+			logrus.WithFields(logFields).Warnf("Grace period exceeded")
+			if err = c.ds.RemoveFinalizerForVolume(vol); err != nil {
+				logFields["error"] = err.Error()
+				logrus.WithFields(logFields).Warnf("Failed to remove finalizer")
+				failed = true
+			} else {
+				logrus.WithFields(logFields).Infof("Removed finalizer")
+			}
+		}
+	}
+	if failed {
+		err = errors.New("Observed error(s) during volume deletion")
+	}
+	return
+}
+
+func (c *UninstallController) deleteEngines(engines map[string]*longhorn.Engine) (err error) {
+	failed := false
+	for _, engine := range engines {
+		logFields := logrus.Fields{
+			"name": engine.Name,
+			"type": "engine",
+		}
+		timeout := metav1.NewTime(time.Now().Add(-GracePeriod))
+		if engine.DeletionTimestamp == nil {
+			if err = c.ds.DeleteEngine(engine.Name); err != nil {
+				logFields["error"] = err.Error()
+				logrus.WithFields(logFields).Warnf("Failed to mark for deletion")
+				failed = true
+			} else {
+				logrus.WithFields(logFields).Infof("Marked for deletion")
+			}
+		} else if engine.DeletionTimestamp.Before(&timeout) {
+			logrus.WithFields(logFields).Warnf("Grace period exceeded")
+			if err = c.ds.RemoveFinalizerForEngine(engine); err != nil {
+				logFields["error"] = err.Error()
+				logrus.WithFields(logFields).Warnf("Failed to remove finalizer")
+				failed = true
+			} else {
+				logrus.WithFields(logFields).Infof("Removed finalizer")
+			}
+		}
+	}
+	if failed {
+		err = errors.New("Observed error(s) during engine deletion")
+	}
+	return
+}
+
+func (c *UninstallController) deleteReplicas(replicas map[string]*longhorn.Replica) (err error) {
+	failed := false
+	for _, replica := range replicas {
+		logFields := logrus.Fields{
+			"name": replica.Name,
+			"type": "replica",
+		}
+		timeout := metav1.NewTime(time.Now().Add(-GracePeriod))
+		if replica.DeletionTimestamp == nil {
+			if err = c.ds.DeleteReplica(replica.Name); err != nil {
+				logFields["error"] = err.Error()
+				logrus.WithFields(logFields).Warnf("Failed to mark for deletion")
+				failed = true
+			} else {
+				logrus.WithFields(logFields).Infof("Marked for deletion")
+			}
+		} else if replica.DeletionTimestamp.Before(&timeout) {
+			logrus.WithFields(logFields).Warnf("Grace period exceeded")
+			if err = c.ds.RemoveFinalizerForReplica(replica); err != nil {
+				logFields["error"] = err.Error()
+				logrus.WithFields(logFields).Warnf("Failed to remove finalizer")
+				failed = true
+			} else {
+				logrus.WithFields(logFields).Infof("Removed finalizer")
+			}
+		}
+	}
+	if failed {
+		err = errors.New("Observed error(s) during replica deletion")
+	}
+	return
+}
+
+func (c *UninstallController) deleteEngineImages(engineImages map[string]*longhorn.EngineImage) (err error) {
+	failed := false
```

Comment:
You can use pattern like this:
https://github.com/rancher/longhorn-manager/blob/master/controller/volume_controller.go#L202